servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/NewRoundRobinLoadBalancer.java

```diff
@@ -250,59 +251,88 @@ public void onNext(@Nullable final Collection<? extends ServiceDiscovererEvent<R
                 LOGGER.debug("{}: unexpectedly received null instead of events.", NewRoundRobinLoadBalancer.this);
                 return;
             }
-            for (ServiceDiscovererEvent<ResolvedAddress> event : events) {
-                final ServiceDiscovererEvent.Status eventStatus = event.status();
-                LOGGER.debug("{}: received new ServiceDiscoverer event {}. Inferred status: {}.",
-                        NewRoundRobinLoadBalancer.this, event, eventStatus);
-
-                @SuppressWarnings("unchecked")
-                final List<Host<ResolvedAddress, C>> usedAddresses =
-                        usedHostsUpdater.updateAndGet(NewRoundRobinLoadBalancer.this, oldHosts -> {
-                            if (isClosedList(oldHosts)) {
-                                return oldHosts;
-                            }
-                            final ResolvedAddress addr = requireNonNull(event.address());
-                            @SuppressWarnings("unchecked")
-                            final List<Host<ResolvedAddress, C>> oldHostsTyped =
-                                    (List<Host<ResolvedAddress, C>>) oldHosts;
-
-                            if (AVAILABLE.equals(eventStatus)) {
-                                return addHostToList(oldHostsTyped, addr);
-                            } else if (EXPIRED.equals(eventStatus)) {
-                                if (oldHostsTyped.isEmpty()) {
-                                    return emptyList();
-                                } else {
-                                    return markHostAsExpired(oldHostsTyped, addr);
-                                }
-                            } else if (UNAVAILABLE.equals(eventStatus)) {
-                                return listWithHostRemoved(oldHostsTyped, host -> {
-                                    boolean match = host.address().equals(addr);
-                                    if (match) {
-                                        host.markClosed();
-                                    }
-                                    return match;
-                                });
-                            } else {
-                                LOGGER.error("{}: Unexpected Status in event:" +
-                                        " {} (mapped to {}). Leaving usedHosts unchanged: {}",
-                                        NewRoundRobinLoadBalancer.this, event, eventStatus, oldHosts);
-                                return oldHosts;
-                            }
-                        });
 
-                LOGGER.debug("{}: now using addresses (size={}): {}.",
-                        NewRoundRobinLoadBalancer.this, usedAddresses.size(), usedAddresses);
+            boolean sendReadyEvent;
+            List<Host<ResolvedAddress, C>> nextHosts = new ArrayList<>();
```

Comment:
1. With new changes, consider going back to line 250 and enhancing it to also check the `events` collection is not empty.
2. Will it make sense to try guessing the array list size vs using the default? Can be smth like a sum of current hosts size + new events size as a worst case scenario to avoid resizing. 
3. Consider marking it as `final`

Reply:
I did 1 and 2, but needed to skip 3 so I can size the list within the loop.